File: DepQBF4Py-0.1/basic-clause-groups-api-example.py
from __future__ import print_function
from DepQBF import *
import logging

#=============================================================================
#  basic-clause-groups-api-example-assumptions.py
#=============================================================================

# The API file 'qdpll.py' has some comments regarding the usage of the API.
# The header file 'qdpll.h' has some comments regarding the usage of the API. 
#
# Please see also the file 'basic-api-example.py' for more
# comments. The example below is similar to 'basic-api-example.py' but
# it does not make use of the push/pop API functions. Instead, clauses
# are added to and deleted from the formula based on selector
# variables. The selector variables are existentially quantified in
# the leftmost quantifier block. Each added clause gets a selector
# variable, which is assigned as an assumption before the actual
# solving starts. This way, clauses are enabled and disabled by
# selector variables. We argue that the use of the push/pop API
# functions is less error-prone from the user's perspective.


# Create solver instance.
qcdcl = QCDCL()

# --dep-man=simple ...  Use the linear ordering of the quantifier prefix.
# --incremental-use ... Enable incremental solving.
qcdcl.configure('--dep-man=simple','--incremental-use')

# Given the following unsatisfiable formula:
#
# p cnf 4 3
# a 1 2 0
# e 3 4 0
# -1 -3 0
# 1 2 4 0
# 1 -4 0
#
# The first clause will be put in one clause group and the last two
# clauses in another group. That last two clauses are unsatisfiable,
# thus deleting the first clause preserves unsatisfiability.


# Declare outermost universal block with variables 1 and 2.
qcdcl.new_scope_at_nesting(QDPLL_QTYPE_FORALL, 1)
qcdcl.add(1)
qcdcl.add(2)
qcdcl.add(0)

# Declare existential block with variables 3 and 4.
qcdcl.new_scope_at_nesting(QDPLL_QTYPE_EXISTS, 2)
qcdcl.add(3)
qcdcl.add(4)
qcdcl.add(0)

# Create a new clause group with ID 'id1'. The ID of a clause group is
# used as its handle and can be passed to API functions.
id1 = qcdcl.new_clause_group()

# Newly created clause groups are closed.
assert (not qcdcl.get_open_clause_group())

# A clause group must be opened before clauses can be added to
# it. Only one clause group can be open at a time.
qcdcl.open_clause_group(id1)
assert (qcdcl.get_open_clause_group () == id1)

# Add the clause '-1 -3 0' to the currently open clause group 'id1'.
qcdcl.add(-1)
qcdcl.add(-3)
qcdcl.add(0)
# The currently open clause group must be closed before creating or
# opening another clause group.
qcdcl.close_clause_group(id1)
assert (not qcdcl.get_open_clause_group())

# Create another clause group 'id2'.
id2 = qcdcl.new_clause_group()
assert (not qcdcl.get_open_clause_group ())
qcdcl.open_clause_group(id2)
assert (qcdcl.get_open_clause_group() == id2)
# Add the clauses '1 2 4 0' and '1 -4 0' to the currently open clause
# group 'id2'.
qcdcl.add(1)
qcdcl.add(2)
qcdcl.add(4)
qcdcl.add(0)
#---------------------
qcdcl.add(1)
qcdcl.add(-4)
qcdcl.add(0)
qcdcl.close_clause_group(id2)
assert (not qcdcl.get_open_clause_group())

qcdcl.print_dimacs()

# Solve the formula, which is unsatisfiable.
res = qcdcl.evaluate()
assert (res == QDPLL_RESULT_UNSAT)
logging.warn('result is %d' % res)
  
# Get a list of those clause groups which contain clauses used by
# solver to determine unsatisfiability. This amounts to an
# unsatisfiable core of the formula.

#check for memory leak
#for i in xrange(100000000):
#    relevant_clause_groups = list(qcdcl.iter_relevant_clause_groups())
#    assert (sum(1 for _ in relevant_clause_groups) == 1)

relevant_clause_groups = list(qcdcl.iter_relevant_clause_groups())

# We must reset the solver AFTER retrieving the relevant groups.
qcdcl.reset ()

# In our example, the clauses in the group 'id2' are relevant for
# unsatisfiability. (The clause '-1 -3 0' in 'id1' cannot be part of a
# resolution refutation found by the solver.)
assert (sum(1 for _ in relevant_clause_groups) == 1)
assert (relevant_clause_groups[0] == id2)

logging.warn('printing zero-terminated relevant clause group IDs:')
print(*relevant_clause_groups, sep='\n')

# Temporarily remove the clause group 'id2' by deactivating it.
logging.warn('deactivating group 2 with clauses 1 2 4 0 and 1 -4 0')
qcdcl.deactivate_clause_group(relevant_clause_groups[0])

# Calling 'qcdcl.gc()' removes superfluous variables and quantifiers
# from the prefix. HOWEVER, in this case, we have deactivated -- not
# deleted -- group 'id2' and hence calling 'qcdcl.gc()' has no effect.
qcdcl.gc()
qcdcl.print_dimacs()

# The formula where group 'id2' has been deactivated is satisfiable.
res = qcdcl.evaluate()
assert (res == QDPLL_RESULT_SAT)
logging.warn('result is %d' % res)
qcdcl.reset()

# Activate group 'id2' again, which makes the formula unsatisfiable.
logging.warn('activating group 2 again')
qcdcl.activate_clause_group(relevant_clause_groups[0])

# Permanently remove the group 'id1'. This operation cannot be undone
# and is in contrast to deactivating a group.
logging.warn('deleting group 1 with clause -1 -3 0')
qcdcl.delete_clause_group(id1)
# Deleting a group invalidates its ID, which can be checked by
# 'qcdcl.exists_clause_group'.
assert (not qcdcl.exists_clause_group(id1))

# Different from the first call of 'qcdcl.gc' above, this time
# variable 3 will be removed from the quantifier prefix. We deleted
# group 'id1' which contains the only clause where variable 3
# occurs. Hence calling 'qcdcl.gc' removes variable 3 because it does
# not occur any more in the formula.
qcdcl.gc()
assert (not qcdcl.is_var_declared(3))
qcdcl.print_dimacs()

# After deleting the group 'id1', the formula consisting only of the
# clauses in group 'id2' is unsatisfiable.
res = qcdcl.evaluate()
assert (res == QDPLL_RESULT_UNSAT)
logging.warn('result is %d\n' %res)

File: examples/basic-clause-groups-api-example.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include "../qdpll.h"

/* Helper function. */
static void
print_relevant_clause_groups (ClauseGroupID * clause_group_ids)
{
  ClauseGroupID *p;
  for (p = clause_group_ids; *p; p++)
    printf ("%d ", *p);
  printf ("0");
}

/* Helper function. */
static unsigned int
count_relevant_clause_groups (ClauseGroupID * clause_group_ids)
{
  unsigned int cnt = 0;
  ClauseGroupID *p;
  for (p = clause_group_ids; *p; p++)
    cnt++;
  return cnt;
}

int
main (int argc, char **argv)
{
  QDPLL *depqbf = qdpll_create ();

  qdpll_configure (depqbf, "--dep-man=simple");
  qdpll_configure (depqbf, "--incremental-use");

  /* Given the following unsatisfiable formula:

     p cnf 4 3
     a 1 2 0
     e 3 4 0
     -1 -3 0
     1 2 4 0
     1 -4 0

     The first clause will be put in one clause group and the last two clauses in
     another group. That last two clauses are unsatisfiable, thus deleting the
     first clause preserves unsatisfiability.
   */

  /* Declare outermost universal block with variables 1 and 2. */
  qdpll_new_scope_at_nesting (depqbf, QDPLL_QTYPE_FORALL, 1);
  qdpll_add (depqbf, 1);
  qdpll_add (depqbf, 2);
  qdpll_add (depqbf, 0);

  /* Declare existential block with variables 3 and 4. */
  qdpll_new_scope_at_nesting (depqbf, QDPLL_QTYPE_EXISTS, 2);
  qdpll_add (depqbf, 3);
  qdpll_add (depqbf, 4);
  qdpll_add (depqbf, 0);

  /* Create a new clause group with ID 'id1'. The ID of a clause group is used
     as its handle and can be passed to API functions. */
  ClauseGroupID id1 = qdpll_new_clause_group (depqbf);
  /* Newly created clause groups are closed. */
  assert (!qdpll_get_open_clause_group (depqbf));
  /* A clause group must be opened before clauses can be added to it. Only one
     clause group can be open at a time. */
  qdpll_open_clause_group (depqbf, id1);
  assert (qdpll_get_open_clause_group (depqbf) == id1);
  /* Add the clause '-1 -3 0' to the currently open clause group 'id1'. */
  qdpll_add (depqbf, -1);
  qdpll_add (depqbf, -3);
  qdpll_add (depqbf, 0);
  /* The currently open clause group must be closed before creating or opening
     another clause group. */
  qdpll_close_clause_group (depqbf, id1);
  assert (!qdpll_get_open_clause_group (depqbf));

  /* Create another clause group 'id2'. */
  ClauseGroupID id2 = qdpll_new_clause_group (depqbf);
  assert (!qdpll_get_open_clause_group (depqbf));
  qdpll_open_clause_group (depqbf, id2);
  assert (qdpll_get_open_clause_group (depqbf) == id2);
  /* Add the clauses '1 2 4 0' and '1 -4 0' to the currently open clause group
     'id2'. */
  qdpll_add (depqbf, 1);
  qdpll_add (depqbf, 2);
  qdpll_add (depqbf, 4);
  qdpll_add (depqbf, 0);
  //---------------------
  qdpll_add (depqbf, 1);
  qdpll_add (depqbf, -4);
  qdpll_add (depqbf, 0);
  qdpll_close_clause_group (depqbf, id2);
  assert (!qdpll_get_open_clause_group (depqbf));

  qdpll_print (depqbf, stdout);

  /* Solve the formula, which is unsatisfiable. */
  QDPLLResult res = qdpll_sat (depqbf);
  assert (res == QDPLL_RESULT_UNSAT);
  printf ("result is %d\n", res);

  /* Get a list of those clause groups which contain clauses used by solver to
     determine unsatisfiability. This amounts to an unsatisfiable core of the
     formula. */
  ClauseGroupID *relevant_clause_groups =
    qdpll_get_relevant_clause_groups (depqbf);
  /* We must reset the solver AFTER retrieving the relevant groups. */
  qdpll_reset (depqbf);
  /* In our example, the clauses in the group 'id2' are relevant for
     unsatisfiability. (The clause '-1 -3 0' in 'id1' cannot be part of a resolution
     refutation found by the solver.) */
  assert (count_relevant_clause_groups (relevant_clause_groups) == 1);
  assert (relevant_clause_groups[0] == id2);
  printf ("printing zero-terminated relevant clause group IDs: ");
  print_relevant_clause_groups (relevant_clause_groups);
  printf ("\n");

  /* Temporarily remove the clause group 'id2' by deactivating it. */
  printf ("deactivating group 2 with clauses 1 2 4 0 and 1 -4 0\n");
  qdpll_deactivate_clause_group (depqbf, relevant_clause_groups[0]);

  /* Calling 'qdpll_gc()' removes superfluous variables and
     quantifiers from the prefix. HOWEVER, in this case, we have
     deactivated -- not deleted -- group 'id2' and hence calling
     'qdpll_gc()' has no effect. */
  qdpll_gc (depqbf);
  qdpll_print (depqbf, stdout);

  /* The formula where group 'id2' has been deactivated is satisfiable. */
  res = qdpll_sat (depqbf);
  assert (res == QDPLL_RESULT_SAT);
  printf ("result is %d\n", res);
  qdpll_reset (depqbf);
  /* Activate group 'id2' again, which makes the formula unsatisfiable. */
  printf ("activating group 2 again\n");
  qdpll_activate_clause_group (depqbf, relevant_clause_groups[0]);

  /* Free memory of array returned by 'qdpll_get_relevant_clause_groups'. 
     This is the caller's responsibility. */
  free (relevant_clause_groups);

  /* Permanently remove the group 'id1'. This operation cannot be undone and
     is in contrast to deactivating a group. */
  printf ("deleting group 1 with clause -1 -3 0\n");
  qdpll_delete_clause_group (depqbf, id1);
  /* Deleting a group invalidates its ID, which can be checked by
     'qdpll_exists_clause_group'. */
  assert (!qdpll_exists_clause_group (depqbf, id1));

  /* Different from the first call of 'qdpll_gc' above, this time variable 3
     will be removed from the quantifier prefix. We deleted group 'id1' which
     contains the only clause where variable 3 occurs. Hence calling 'qdpll_gc'
     removes variable 3 because it does not occur any more in the formula. */
  qdpll_gc (depqbf);
  assert (!qdpll_is_var_declared (depqbf, 3));
  qdpll_print (depqbf, stdout);

  /* After deleting the group 'id1', the formula consisting only of the
     clauses in group 'id2' is unsatisfiable. */
  res = qdpll_sat (depqbf);
  assert (res == QDPLL_RESULT_UNSAT);
  printf ("result is %d\n", res);

  qdpll_delete (depqbf);
}
